### PR TITLE
Disallow using In/Out structs with From/As annotations

### LINKED ISSUE #1060: Disallow using In/Out structs with From/As annotations
In #1053, we eased up the requirement for function signature for applying annotations to allow using In/Out structs with ResultTags/ParamTags. This had an unintended change in behavior where users are now able to use In/Out structs with From/As annotations. For example, the below code used to error out:
```
type Foo interface { ... }
type Bar struct { ... } // Bar implements Foo
struct Res {
  fx.Out

  ABar Bar
}
fx.Annotate(
  func() Res { ... },
  fx.As(new(Foo)),
)
```
Currently the code above applies the As annotation to `ABar` field and transforms the target function's signature accordingly. However, this was an unintended change in behavior, and we should have a design discussion about whether we actually want this before cutting a new release. Since it will require a breaking change to disable this after a new release is cut, I think the safe approach is to disable this for now and enable it after having decided that this is indeed what we want.

### PR DESCRIPTION
**Pull request by ankit209**
In https://github.com/uber-go/fx/pull/1053, we eased up the requirement for function signature for applying annotations to allow using In/Out structs with ResultTags/ParamTags. This had an unintended change in behavior where users are now able to use In/Out structs with From/As annotations. For example, the below code used to error out:

> type Foo interface { ... }
> type Bar struct { ... } // Bar implements Foo
> struct Res {
>   fx.Out
> 
>   ABar Bar
> }
> fx.Annotate(
>   func() Res { ... },
>   fx.As(new(Foo)),
> )

Currently the code above applies the As annotation to ABar field and transforms the target function's signature accordingly. However, this was an unintended change in behavior, and we should have a design discussion about whether we actually want this before cutting a new release. Since it will require a breaking change to disable this after a new release is cut, I think the safe approach is to disable this for now and enable it after having decided that this is indeed what we want.

Fixes #1060

### CODE REVIEW

**Comment by JacobOaks** [comment]
I understand disallowing `fx.From` with `fx.In` and `fx.Out` with `fx.As`. But is there any reason why we should be disallowing `fx.From` (applies to inputs only) with a function that returns an `fx.Out` (is an output) and `fx.As` (applies to outputs only) with a function that returns an `fx.In` (is an input)?

File: annotated.go
Comment (tchung1118) [inline annotated.go:1620]: This is actually the old behavior we had for As/From annotations with Out/In structs, which is the safest way to go before easing up on this requirement. However, I agree with Jacob that this is a little overly restrictive, considering As annotation does not affect parameters of a constructor, and From annotation does not affect results of a constructor at all. So we could probably let users use As and From annotations with In and Out structs, respectively.

I'll leave it up to you to decide on how you'd like to proceed.

File: annotated_test.go
Comment (tchung1118) [inline annotated_test.go:1454]: Depending on what you do with my first comment, you might want to add more test cases. Specifically, with the current implementation, you're missing test cases using fx.From with fx.Out and fx.As with fx.In. If you decide to let users be able to use those cases, you don't need to add them to tests.

File: annotated.go
Comment (tchung1118) [inline annotated.go:1623]: Extraneous empty line

**Review by tchung1118 — COMMENTED** [review]
Leaving couple of comments. Also lint seems to have failed? (try running `make lint`)

File: annotated.go
Comment (ankit209) [reply annotated.go]: Makes sense. 
Have removed this unintended restriction and to allow fx.Out struct with From annotation and fx.In struct with As annotation.

**Comment by ankit209** [comment]
> I understand disallowing `fx.From` with `fx.In` and `fx.Out` with `fx.As`. But is there any reason why we should be disallowing `fx.From` (applies to inputs only) with a function that returns an `fx.Out` (is an output) and `fx.As` (applies to outputs only) with a function that accepts an `fx.In` (is an input)?

Agreed Jacob. Thanks for calling this out, I have made the necessary edits.

**Comment by ankit209** [comment]
> Leaving couple of comments. Also lint seems to have failed? (try running `make lint`)

Thanks @tchung1118. Have made the required edits and fixed lint errors as well.

File: annotated_test.go
Comment (JacobOaks) [inline annotated_test.go:1460]:
```suggestion
			// implements I
```

**Review by JacobOaks — COMMENTED** [review]
This LGTM but I'll wait to see if @tchung1118 has any outstanding comments.

**Review by tchung1118 — APPROVED** [review]
LGTM

File: annotated.go
Comment (sywhang) [inline annotated.go:1328]:
```suggestion
// From annotation cannot be used in a function that takes an [In] struct as a
```

File: annotated.go
Comment (sywhang) [inline annotated.go:1599]:
```suggestion
// - returns an fx.Out struct as a result and has either a ResultTags or an As annotation
```

File: annotated.go
Comment (sywhang) [inline annotated.go:1620]:
```suggestion
			return errors.New("fx.Out structs cannot be annotated with fx.ResultTags or fx.As")
```

File: annotated.go
Comment (sywhang) [inline annotated.go:1632]:
```suggestion
			return errors.New("fx.In structs cannot be annotated with fx.ParamTags or fx.From")
```

File: annotated_test.go
Comment (sywhang) [inline annotated_test.go:1481]:
```suggestion
		assert.Contains(t, err.Error(), "fx.Out structs cannot be annotated with fx.ResultTags or fx.As")
```

File: annotated_test.go
Comment (sywhang) [inline annotated_test.go:1529]:
```suggestion
		assert.Contains(t, err.Error(), "fx.In structs cannot be annotated with fx.ParamTags or fx.From")
```

**Review by sywhang — APPROVED** [review]

**Comment by ankit209** [comment]
Thanks for correcting the documentation and approving the PR @sywhang

**Pull request merged**
